File: scripts/reporter.py
import subprocess
import os
import sys

REPORTER_PS1 = os.path.join(os.path.dirname(os.path.abspath(__file__)), "windows-reporter.ps1")

def report_status(status: str, title: str, message: str = "", percent: int = -1, timeout: int = 5, async_mode: bool = True) -> bool:
    """
    Send a Windows Report Popup status update for AI Commander jobs.
    Statuses: RUNNING, PROGRESS, COMPLETED, ERROR
    """
    cmd = [
        "powershell.exe",
        "-NoProfile",
        "-ExecutionPolicy", "Bypass",
        "-File", REPORTER_PS1,
        "-Status", status,
        "-Title", title,
        "-Message", message,
        "-TimeoutSeconds", str(timeout)
    ]
    if percent >= 0:
        cmd.extend(["-Percent", str(percent)])
    if async_mode:
        cmd.append("-Async")
        
    try:
        res = subprocess.run(cmd, capture_output=True, text=True, timeout=10)
        return res.returncode == 0
    except Exception as e:
        print(f"[ReporterError] Failed to trigger popup: {e}", file=sys.stderr)
        return False

class WorkflowReporter:
    """
    Context manager and helper class for AI Commander workflow jobs.
    
    Usage:
        with WorkflowReporter("Data Sync Job", "Starting sync...") as r:
            r.progress(35, "Fetching records...")
            # do work...
            r.progress(80, "Writing database...")
            # finished automatically with COMPLETED popup
    """
    def __init__(self, title: str, initial_msg: str = "Task started...", timeout: int = 4):
        self.title = title
        self.timeout = timeout
        report_status("RUNNING", self.title, initial_msg, timeout=self.timeout, async_mode=True)

    def progress(self, percent: int, msg: str = ""):
        report_status("PROGRESS", self.title, msg, percent=percent, timeout=self.timeout, async_mode=True)

    def completed(self, msg: str = "Task completed successfully.", timeout: int = 5):
        report_status("COMPLETED", self.title, msg, percent=100, timeout=timeout, async_mode=True)

    def error(self, msg: str = "Task encountered an error.", timeout: int = 8):
        report_status("ERROR", self.title, msg, timeout=timeout, async_mode=True)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_type is not None:
            self.error(f"{exc_type.__name__}: {exc_val}")
        else:
            self.completed()
        return False

File: scripts/reporter.js
import { spawn } from 'node:child_process';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPORTER_PS1 = path.join(__dirname, 'windows-reporter.ps1');

export function reportStatus(status, title, message = '', percent = -1, timeoutSeconds = 5, asyncMode = true) {
  const args = [
    '-NoProfile',
    '-ExecutionPolicy', 'Bypass',
    '-File', REPORTER_PS1,
    '-Status', status,
    '-Title', title,
    '-Message', message,
    '-TimeoutSeconds', String(timeoutSeconds)
  ];
  if (percent >= 0) {
    args.push('-Percent', String(percent));
  }
  if (asyncMode) {
    args.push('-Async');
  }

  try {
    const child = spawn('powershell.exe', args, { detached: true, stdio: 'ignore' });
    child.unref();
    return true;
  } catch (err) {
    console.error('[ReporterError] Failed to spawn popup:', err);
    return false;
  }
}

export class WorkflowReporter {
  constructor(title, initialMsg = 'Task started...', timeout = 4) {
    this.title = title;
    this.timeout = timeout;
    reportStatus('RUNNING', this.title, initialMsg, -1, this.timeout, true);
  }

  progress(percent, msg = '') {
    reportStatus('PROGRESS', this.title, msg, percent, this.timeout, true);
  }

  completed(msg = 'Task completed successfully.', timeout = 5) {
    reportStatus('COMPLETED', this.title, msg, 100, timeout, true);
  }

  error(msg = 'Task encountered an error.', timeout = 8) {
    reportStatus('ERROR', this.title, msg, -1, timeout, true);
  }
}
